pinot-common/src/main/java/com/linkedin/pinot/common/utils/time/DateTimeFieldSpecUtils.java

```diff
@@ -0,0 +1,393 @@
+   * @param format
+   * @return
+   */
+  public static boolean validFormat(String format) {
+
+    Preconditions.checkNotNull(format);
+    String[] formatTokens = format.split(COLON_SEPARATOR);
+    Preconditions.checkArgument(formatTokens.length == MIN_FORMAT_TOKENS || formatTokens.length == MAX_FORMAT_TOKENS,
+        FORMAT_TOKENS_ERROR_STR);
+    Preconditions.checkArgument(formatTokens[FORMAT_SIZE_POSITION].matches(NUMBER_REGEX)
+        && EnumUtils.isValidEnum(TimeUnit.class, formatTokens[FORMAT_UNIT_POSITION]), FORMAT_PATTERN_ERROR_STR);
+    if (formatTokens.length == MIN_FORMAT_TOKENS) {
+      Preconditions.checkArgument(formatTokens[FORMAT_TIMEFORMAT_POSITION].equals(TimeFormat.EPOCH.toString()),
+          TIME_FORMAT_ERROR_STR);
+    } else {
+      Preconditions.checkArgument(formatTokens[FORMAT_TIMEFORMAT_POSITION].equals(TimeFormat.SIMPLE_DATE_FORMAT.toString()),
+          TIME_FORMAT_ERROR_STR);
+      Preconditions.checkNotNull(formatTokens[FORMAT_PATTERN_POSITION]);
+    }
+    return true;
+  }
+
+  /**
+   * Check correctness of granularity of {@link DateTimeFieldSpec}
+   * @param granularity
+   * @return
+   */
+  public static boolean validGranularity(String granularity) {
+    Preconditions.checkNotNull(granularity);
+    String[] granularityTokens = granularity.split(COLON_SEPARATOR);
+    Preconditions.checkArgument(granularityTokens.length == MAX_GRANULARITY_TOKENS, GRANULARITY_TOKENS_ERROR_STR);
+    Preconditions.checkArgument(granularityTokens[GRANULARITY_SIZE_POSITION].matches(NUMBER_REGEX)
+        && EnumUtils.isValidEnum(TimeUnit.class, granularityTokens[GRANULARITY_UNIT_POSITION]), GRANULARITY_PATTERN_ERROR_STR);
+
+    return true;
+  }
+
+  /**
+   * Given a timestamp in millis, convert it to the format specified, while also handling time units such as WEEKS, MONTHS, YEARS
+   * This is a wrapper over the fromMillisToFormat method, which only accepts units in TimeUnits
+   * eg:
+   * 1) given dateTimeColumnValueMS = 1498892400000 and format=1:HOURS:EPOCH,
+   * dateTimeSpec.fromMillis(1498892400000) = 416359 (i.e. dateTimeColumnValueMS/(1000*60*60))
+   *
+   * 2) given dateTimeColumnValueMS = 1498892400000 and format=1:WEEKS:EPOCH,
+   * dateTimeSpec.fromMillis(1498892400000) = 2478 (i.e. timeColumnValueMS/(1000*60*60*24*7))
+   *
+   * 3) given dateTimeColumnValueMS = 1498892400000 and format=1:DAYS:SIMPLE_DATE_FORMAT:yyyyMMdd,
+   * dateTimeSpec.fromMillis(1498892400000) = 20170701
+   * @param dateTimeValueMillis - date time value in millis to convert
+   * @param outputFormat
+   * @return
+   */
+  public static Long convertMillisToCustomFormat(Long dateTimeValueMillis, String outputFormat) {
+    Preconditions.checkNotNull(dateTimeValueMillis);
+    Preconditions.checkNotNull(outputFormat);
+
+    Long convertedDateTime = null;
+    String timeUnit = outputFormat.split(COLON_SEPARATOR)[FORMAT_UNIT_POSITION];
+
+    if (EnumUtils.isValidEnum(TimeUnit.class, timeUnit)) {
```

Comment:
Having a single converter that handles all time units will avoid the if/else here.

Reply:
fixed